File: 287-find-the-duplicate-number/287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        
        int n = nums.length;
        int freq[] = new int[n];
        
        System.out.println(Arrays.toString(freq));
        
        for(int i=0; i<nums.length; i++){
            if(freq[nums[i]] == 0){
                freq[nums[i]] = 1;
            }else{
                return nums[i];
            }
        }
        
        return -1;
    }
}

File: 287-find-the-duplicate-number/287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        
        int n = nums.size();
        
        int slow = 0;
        int fast = 0;
        for(int i=0; i<n; i++){
        
            
            if(slow == 0 && fast ==0){
                slow = nums[slow];
                fast = nums[nums[fast]];
            }
            else if(slow != fast){
                slow = nums[slow];
                fast = nums[nums[fast]];
            }
            else{
                fast = 0;
                while(fast != slow){
                    slow = nums[slow];
                    fast = nums[fast];
                }
                return fast;
            }
        }
        
    return -1;
    }
};